Recursive stage2 parser

hryx opened the pull request:
## TODO

- [x] :trophy: Finish implementation such that CI passes
- [x] :boxing_glove: Remove as many `TODO` comments as possible by getting answers to questions in code review and creating follow-ups and tickets as needed
- [x] :speech_balloon: Address all review comments
- [x] :shower: Squish several commits via interactive rebase to clean up history a bit
- [x] :clipboard: Fill out remaining sections of this pull request description

---

Closes #1729

Keep your eyes open for any `TODO` comments and leave your suggestions or opinions. Some of these comments will likely be retained and their corresponding changes postponed till a later PR, but maybe some of them can be solved now.

## Performance impact

Here are some [informal tests of `parser_test.zig` with `perf stat -d`](https://gist.github.com/hryx/91a656944c55039139391939a052d990) on x86_64 Linux. In the absence of visualizations and more formal testing, some quick findings based on my system:
- CPU cycles, instructions, branches: -55%
- Page faults are roughly the same
- Real time elapsed: -45%

## Readability impact

Indentation stats of `std/zig/parse.zig`:

```
master                          | stage2-recursive-parser
--------------------------------+---------------------------------
indent count                    | indent count
------------                    | ------------
    0     92                    |      0   263
    1    241                    |      1   803
    2    123                    |      2   763
    3    291                    |      3   334
    4    654                    |      4   133
    5    662                    |      5    60
    6    700                    |      6    28
    7    347                    |      7     5
    8    229                    |
    9     42                    |
   10     18                    |
avg indentation level: 4.796999 | avg indentation level: 1.827543
source lines of code:      3399 | source lines of code:      2389
```

- Maximum indentation level went from 10 to 7
- Average indentation level went from 4.79 to 1.82
- Lines of code (excluding blank/comments) went from 3399 to 2389

## Follow-up items

Because this was a heavy change, a few ideas and concerns came up along the way which would unnecessarily increase the scope of this PR or add diff noise but which I would like to address.

- Better data modeling of VarDecl initialization: https://github.com/ziglang/zig/pull/2405#discussion_r283098325
- Fix CompareExpr in stage1: https://github.com/ziglang/zig/pull/2405/files#r282849239
- `ArrayTypeStart` can be either an array or a slice, but `const` and `allowzero` only work on pointers; slices are pointers but arrays aren't. The `PrefixTypeOp` rule might need to be updated to prevent those qualifiers on slices, e.g.:
```diff
 PrefixTypeOp
     <- QUESTIONMARK
      / KEYWORD_promise MINUSRARROW
-     / ArrayTypeStart (ByteAlign / KEYWORD_const / KEYWORD_volatile / KEYWORD_allowzero)*
+     / ArrayTypeStart
+     / SliceTypeStart (ByteAlign / KEYWORD_const / KEYWORD_volatile / KEYWORD_allowzero)*
      / PtrTypeStart (KEYWORD_align LPAREN Expr (COLON INTEGER COLON INTEGER)? RPAREN / KEYWORD_const / KEYWORD_volatile / KEYWORD_allowzero)*
```
- Implement disallowing duplicate pointer qualifiers in stage1. Add this test to `test/compile_errors.zig`:
```zig
cases.add(
    "duplicate qualifiers",
    \\test "duplicates" {
    \\    var a: *align(4) align(4) i32 = 0;
    \\    var b: *const const i32 = 0;
    \\    var c: *volatile volatile i32 = 0;
    \\    var d: *allowzero allowzero i32 = 0;
    \\}
,
    "tmp.zig:2:22: error: Extra align qualifier",
    "tmp.zig:3:19: error: Extra const qualifier",
    "tmp.zig:4:22: error: Extra volatile qualifier",
    "tmp.zig:5:23: error: Extra allowzero qualifier",
);
```
- The existing parse error system is somewhat cumbersome: Every possible error is its own type, has entries in the switches for the `render()` and `loc()` methods, and exists in the `ast.Error` union. If it's possible, it would be more flexible, concise, legible, and expressive to generate errors from string literals at the site of the error itself. Something like this:
```zig
const expr_node = try expectNode(arena, it, tree, parseExpr, "expression or assignment");
```

## Miscellany

:bowing_man: Massive thanks to @tyler569 for unblocking me by finding a memory leak I caused along the way.

:rose: I am so sorry I killed your art @Hejsil

tiehuis commented on std/zig/parse2.zig:
```diff
@@ -0,0 +1,2978 @@
+    node.* = Node.Root{
+        .base = Node{ .id = .Root },
+        .decls = undefined,
+        // TODO: File an issue: Currently it is impossible to have a container-level
```
Believe this is covered by #2288?

hryx replied:
Ah, I think you're right @tiehuis -- although it's a variation on the problem I encountered. I tried a top-level doc comment with _no_ comment on the first item, rather than your example of both comments becoming collapsed into one. I think they are both caused by the same problem: `zig fmt` collapses newlines below comments. It seems to me that behavior will have to change in order to fix the problem -- what do you think?

(As an aside, I have bigger changes in mind for doc comments in general, but that will definitely have to wait until the parser rewrite is finished.)

Hejsil commented:
Exciting. I'll look through all of this once you're done :)

andrewrk commented:
All tests passing :tada: :tada: 

Ready for some review?

hryx commented:
I still have some work to do to finish the PR description, but since this a big one, I suppose it's time to get started. Let's do it!

hryx commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-/// Result should be freed with tree.deinit() when there are
-/// no more references to any of the tokens or nodes.
-pub fn parse(allocator: *mem.Allocator, source: []const u8) !*ast.Tree {
+pub const Error = error{UnexpectedToken} || Allocator.Error;
```
`UnexpectedToken` is probably the wrong name for this error. Probably `ParseError` instead. (This error is never returned from `pub fn parse`.)

hryx commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+
+    if (try parseVarDecl(arena, it, tree)) |node| {
+        var var_decl = node.cast(Node.VarDecl).?;
+        var_decl.*.doc_comments = doc_comments;
```
A few `ptr.*.field`-style accesses pepper this PR, where otherwise it uses the `ptr.field` implicit dereferencing. I wasn't sure which is more idiomatic (I'm guessing implicit); it should probably be made consistent.

emekoi replied:
we should probably, put something about that in the style guide.

hryx replied:
:+1: I'll not worry about it for now and wait until there is a style guideline for it.

andrewrk commented:
>     * This branch has too many commits. How should the git history be cleaned up when ready to merge?

For a branch this large, if there are conflicts, merge master into the branch, then fix the conflicts. If there are no conflicts - which it looks to be the case - no work is needed on your part, I'll merge it into master with a merge commit.

andrewrk commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+    };
+    node.decls = parseContainerMembers(arena, it, tree, .Keyword_struct) catch |err| {
+        // TODO: Switch on the error type
+        // https://github.com/ziglang/zig/issues/2203
```
I went ahead and solved #2203 so that you could remove this TODO comment.

On the other hand, maybe this should be a new proposal because I see here you have this idea that you could have an entire error set as one of the switch prongs. That's a great idea!

hryx replied:
Done! https://github.com/ziglang/zig/issues/2473

andrewrk commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+    return node;
+}
+
+// ContainerMembers
```
This is not super important but - I think you can go ahead and make these doc comments. I envision documentation generation not only as a way to communicate the external API of a library, but also as a way to browse and explore a codebase.

andrewrk commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
 
-    try stack.append(State.TopLevel);
+        if (try parseTestDecl(arena, it, tree)) |node| {
+            node.cast(Node.TestDecl).?.doc_comments = doc_comments;
```
Why not avoid this cast by having `parseTestDecl` return `*Node.TestDecl`? I see this pattern in a few other places.

hryx replied:
I agree it would be nice to avoid the casts. We could probably do this for a bunch of these rule parsers. However I think this might conflict (philosophically) with what @Hejsil requested [over here](https://github.com/ziglang/zig/pull/2405#discussion_r282834107) when one of the rule parsers deviated from the common signature: 

> We should strive to have all functions follow the ParseFn signature when possible

I might be reading too far into it, of course. Anyway, if there's any conflict of opinion I think I'll leave it as-is for now and revisit another time, if that's cool!

hryx replied:
(Of course, I should note that by necessity, all of the rule parsers that return a `Node.SomeKindaList` type already deviate from this signature by not returning a `*Node`. FWIW.)

daurnimator commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+pub fn parse(allocator: *Allocator, source: []const u8) !*Tree {
     var tree_arena = std.heap.ArenaAllocator.init(allocator);
     errdefer tree_arena.deinit();
+    const arena = &tree_arena.allocator;
```
This should be refactored similar to https://github.com/ziglang/zig/pull/2424/commits/cb4fb14b6e66bd213575f69eec9598be8394fae6

hryx replied:
Note to self: don't copy `tree_arena`

andrewrk replied:
Yeah there's a bit of a footgun here - sorry about that. The in-progress copy elision changes + "fixed" fields should be able to address this.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
`PrimaryTypeExpr` parses `ErrorSetDecl` before `KEYWORD_error DOT IDENTIFIER`. If this was `expectToken` then `KEYWORD_error DOT IDENTIFIER` would never be reached.

Hejsil replied:
The stage 1 parses errors here instead of putting back the tokens. I can't remember the exact reasoning, but I do think erroring here is correct until someone demonstrates otherwise.

Hejsil replied:
`parseLoopTypeExpr` have this piece of code.
```zig
    // If we've seen "inline", there should have been a "for" or "while"
    try tree.errors.push(AstError{
        .ExpectedInlinable = AstError.ExpectedInlinable{ .token = it.index },
    });
```

Why doesn't `parseLoopExpr` do the same?

Hejsil replied:
This should probably also return a function instead of a `type`.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
I think it would be better if this function returned `ParseFn` instead of `type`.

hryx replied:
:+1: In order to accommodate `ListParser`, which needs to return `!SomeListType` instead of `!?*Node` like the others, I've made `ParseFn` a generic which returns a function type. So now most parse functions are of type `ParseFn(?*Node)` instead of type `ParseFn`. Hope that makes sense.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
Shouldn't `op_token` be assigned to `token`?

hryx replied:
Very derpy on my behalf. Great catch. :>

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
I don't think we should encode in the grammar that only one `volatile` (or any other modifier) should be allowed on pointers. It would make the grammar more complicated than it has to be.

However, I think it should be an error. Grammars do not enforce semantics, and we can define that it is semantically illegal to have 2 `const` or whatever. Where we catch this semantic error is up to the implementation. I think the easiest place to catch it is probably in the parser in this function.

hryx replied:
Ok, I see what you mean. Originally I was thinking that the rule should be more like this:

    / ArrayTypeStart ByteAlign? KEYWORD_const? KEYWORD_volatile? KEYWORD_allowzero?

But that would enforce ordering, which would be a bit like prescribing an ordering to prefix operators. Which would be poopy.

I'll implement your suggestion of emitting an error. That seems especially sound given that multiple align qualifiers could conflict.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
Hmmm. Yeah maybe these should have different rules.

hryx replied:
Rad, I'll leave this TODO here for a future fix.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
You could define a local struct to this function and use that:
```zig
const OpAndToken = struct { op: Node.SuffixOp.Op, token: TokenIndex };
const op_and_token = blk: { ...
```

Ofc, anonymous structs would be better.

hryx replied:
Brilliant idea. Bless this damn language.

Hejsil commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
Hmm. This comment is showing the wrong line of code for me here. This should be on this line:
```zig
fn prevToken(it: *TokenIterator) void {
```

Hejsil replied:
Hmm. Seems that a lot of my comments are showing this line for some reason... If i go to "Files changed", the comments are in the right place.

hryx replied:
Good idea. Additionally, I might use this opportunity to rename it to `putBackToken`, since the purpose of this function is no longer to "get the previous token".

(Don't worry, I see your comments in the proper places under "Files changed". Maybe it's an artifact of the ridiculous/incomprehensible diff to `parse.zig`.)

hryx replied:
This ended up looking nice, and quite a bit more self-documenting!

andrewrk commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+        .visib_token = null,
+        .thread_local_token = null,
+        .name_token = name_token,
+        .eq_token = eq_token orelse 0,
```
Oops I missed this. Yes, this should probably be, above, `const eq_token = eatToken(it, .Equal) orelse (return a parse error)`

hryx replied:
Good catch, I completely forgot to leave a `TODO` -- and still getting used to treating `undefined` as a debug tool.

I see what happened here. I didn't use `expectToken` because the grammar shows the assignment is optional: `(EQUAL Expr)?`. Presumably this is to support externs, which allow variable decls with no initialization (`extern const food: SomeType;`), unlike regular variables. I'd grepped the repo for such statements and found nothing, but now I get it.

Between making `eq_token` optional and leaving it undefined, in this case I think I'd prefer the latter. Less impact on other code, and it's only accessed when `init_node` is present: https://github.com/ziglang/zig/blob/master/std/zig/render.zig#L1801-L1804

andrewrk replied:
Ahh, I apologize, I misunderstood and gave bad advice. I see now the issue, and I think your proposed solution is fine. The other, more destructive but perhaps better data modeling approach, would be:

```diff
--- a/std/zig/ast.zig
+++ b/std/zig/ast.zig
@@ -508,7 +508,6 @@ pub const Node = struct {
         visib_token: ?TokenIndex,
         thread_local_token: ?TokenIndex,
         name_token: TokenIndex,
-        eq_token: TokenIndex,
         mut_token: TokenIndex,
         comptime_token: ?TokenIndex,
         extern_export_token: ?TokenIndex,
@@ -516,9 +515,14 @@ pub const Node = struct {
         type_node: ?*Node,
         align_node: ?*Node,
         section_node: ?*Node,
-        init_node: ?*Node,
+        init: ?Init,
         semicolon_token: TokenIndex,
 
+        pub const Init = struct {
+            eq_token: TokenIndex,
+            expr: *Node,
+        };
+
```

I think your approach sounds good, we can do another pass on the ast structures later and see if we want to do these kinds of data modeling improvements.

hryx commented:
Thanks for the review and great comments @Hejsil & @andrewrk. Been a busy week but I will have time to address all these comments this weekend. Looking forward to wrapping up!

Sahnvour commented:
>     * This branch has too many commits. How should the git history be cleaned up when ready to merge?

Looking through the commits, I think you can first reduce their number by doing an interactive rebase and use many small commits as fixups for previous ones.

hryx commented:
Thanks @Sahnvour, that's my plan now. After addressing all these review comments today, I'll turn a bunch of commits into fixups. But since it sounds like having several commits is allowed, I won't obsess over getting it perfect.

hryx commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                continue;
-            },
-            State.LabeledExpression => |ctx| {
-            
```
:+1: I'll add this to the follow-up list.

hryx commented:
Those :x: marks gave me a good scare; all tests pass locally. I assume Azure cancels a run when a newer commit is pushed on the branch.

hryx commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
-                    }) catch unreachable;
-                    continue;
-                }
+    switch (kind) {
```
I'm gonna go ahead and implement your suggestion, partly for the cleanup and partly because it actually conforms better to the grammar. Fortunately (?), none of these are implemented in `ir.zig` yet, so the impact on outside code is negligible.

hryx commented on std/zig/parse.zig:
```diff
@@ -1,3659 +1,3044 @@
+        .ExpectedReturnType = AstError.ExpectedReturnType{ .token = it.index },
+    });
+
+    // TODO: Based on this rule, `!var` is an acceptable return type, but there is no usage
```
I see now, this is https://github.com/ziglang/zig/issues/447 (Note that that proposal is marked accepted.) For some reason I thought this was an existing feature.

I'll extract this TODO grammar idea and just leave a comment on that ticket in case it's useful.

hryx commented:
The git history has been cleaned up. Please don't ask me how I did it. The original 138 commits can be seen at this tag: https://github.com/ziglang/zig/compare/master...hryx:stage2-recursive-parser-pre-cleanup

Regarding the most recent changes: some other files did get touched, either to address review comments or to make the parser conform more strictly to the grammar rules. Either way, I now consider this **ready to merge** if there are no more concerns or blockers!

and now, art

:white_large_square: :white_large_square: :white_large_square: :white_large_square: :sun_with_face: :white_large_square: :white_large_square: :white_large_square:
:white_large_square: :white_large_square: :white_large_square: :white_large_square: :white_large_square: :white_large_square: :white_large_square: :white_large_square:
:white_large_square: :santa: :sweat_drops: :white_large_square: :white_large_square:  :white_large_square: :white_large_square: :boat: 
:ocean: :ocean: :ocean: :ocean: :ocean: :shark: :ocean: :ocean:

daurnimator commented on std/zig.zig:
```diff
@@ -2,6 +2,7 @@ const tokenizer = @import("zig/tokenizer.zig");
 pub const Token = tokenizer.Token;
 pub const Tokenizer = tokenizer.Tokenizer;
 pub const parse = @import("zig/parse.zig").parse;
+pub const parse2 = @import("zig/parse2.zig").parse;
```
Is this intentional?

hryx replied:
Nope, artifact of the funky squish 'n' rebase. I'll fix.

daurnimator commented on std/zig/ast.zig:
```diff
@@ -182,25 +238,61 @@ pub const Error = union(enum) {
+    pub const ExpectedLoopExpr = SingleTokenError("Expected loop expression, found {}");
+    pub const ExpectedDerefOrUnwrap = SingleTokenError("Expected pointer dereference or optional unwrap, found {}");
+    pub const ExpectedSuffixOp = SingleTokenError("Expected pointer dereference, optional unwrap, or field access, found {}");
+
```
Empty line

hryx replied:
There was already a blank newline separating the `SingleTokenError` declarations from the `SimpleError` declarations.

daurnimator commented on std/zig/render.zig:
```diff
@@ -1668,15 +1634,15 @@ fn renderExpression(
             var it = asm_node.clobbers.iterator(0);
             while (true) {
-                const clobber_token = it.next().?;
+                const clobber_node = it.next().?.*;
```
Can there be 0 items?

hryx replied:
Nothing changed here. Length of `clobbers` is already checked here: https://github.com/ziglang/zig/blob/0d629421c97f5be7a89102c0ae57916418e71243/std/zig/render.zig#L1617-L1620

But now that you bring my attention here, it would be nicer if the section below matched the `if-payload` style of the other asm parts, instead of `if (it.peek() == null)`: https://github.com/ziglang/zig/blob/0d629421c97f5be7a89102c0ae57916418e71243/std/zig/render.zig#L1608
